importmap: setup importmap-rails

otegami opened the pull request:
GitHub: ref GH-39

This PR migrated the setup from `Webpacker` to `Importmap`. Now we can handle JavaScripts using Importmap.
- ref: https://github.com/rails/importmap-rails

## What we will do

- [x] Add importmap-rails
- [x] Set up and configure Importmap. <- This PR is here.
- [ ] Remove the Webpacker dependency.

otegami committed:
rails importmap:install

otegami committed:
importmap: install existed packages

- bin/importmap pin @rails/actioncable
- bin/importmap pin @rails/ujs@7.0.8-4

otegami committed:
Replace `javascript_pack_tag` with `javascript_importmap_tags`

Because we will use Importmap instead of Webpacker.

otegami committed:
Move an entry point from 'app/javascript/packs' to 'app/javascript'

otegami committed:
Rewrite channels/index.js in ES6 format

otegami commented on app/javascript/channels/index.js:
```diff
@@ -1,5 +1 @@
-
-const channels = require.context('.', true, /_channel\.js$/)
-channels.keys().forEach(channels)
+// Import all the channels to be used by Action Cable
```
This is current default setting in Rails.
If we want to import the channels, we will do it.
In this moment,  we don't have any channels to import.

- https://github.com/rails/rails/blob/7-0-stable/actioncable/lib/rails/generators/channel/templates/javascript/index.js.tt

otegami commented on app/javascript/application.js:
```diff
@@ -0,0 +1,5 @@
+import Rails from "@rails/ujs"
+import "channels"
+
+Rails.start()
```
Entry point is moved from `app/javascript/packs/application.js` to `app/javascript/application.js`.
When using Webpacker, the entry point is under packs.

otegami commented on bin/importmap:
Auto generated files.

otegami commented on vendor/javascript/@rails--ujs.js:
This minified javascript is automatically added using `bin/importmap pin @rails/ujs@7.0.8-4` as a default.
- ref: https://ga.jspm.io/npm:@rails/ujs@7.0.8-4/lib/assets/compiled/rails-ujs.js

otegami commented:
## How I confirmed the change

Add the following change at `app/javascript/application.js`.
- This is the entry point of JavaScript now.

```diff
$ git diff
diff --git a/app/javascript/application.js b/app/javascript/application.js
index 077f320..7a3084d 100644
--- a/app/javascript/application.js
+++ b/app/javascript/application.js
@@ -2,4 +2,6 @@
 import Rails from "@rails/ujs"
 import "channels"

+console.log("Importmap works in Ranguba!")
+
 Rails.start()
```

Start Rails Server

```console
$ bin/rails s
```

Check whether the message appear or not in the console log as follow.

![Screenshot from 2024-07-26 09-33-34](https://github.com/user-attachments/assets/fb70856e-d69b-48a7-b517-75a49b78a65c)

abetomo commented on vendor/javascript/.keep:
Do we need this file?

otegami replied:
fix: [67f8a22](https://github.com/ranguba/ranguba/pull/41/commits/67f8a22582425d9f7543721d1fea08190128ff6c) Thanks, I just deleted it because using this file just kept this `vendor/javascript` directory using git.

otegami committed:
Remove an unnecessary file

Because there is a file under this directory now.
We don't have to keep it to manage it by git.

kou commented on vendor/javascript/@rails--ujs.js:
Do we need to add this file to our repository? (Is it the standard way?)
Can we generate this automatically when we deploy our application?

otegami replied:
> Do we need to add this file to our repository? (Is it the standard way?)

By default, packages are downloaded using `bin/importmap pin package-name` as follows although packages were optionaly downloaded in the past like [here](https://github.com/rails/importmap-rails/pull/217).
So I think this is the standard way now, although I don't have strong confidence.

https://github.com/rails/importmap-rails/blob/0a6330133b75a72d7c5d07fe87c2fce50d76b997/lib/importmap/commands.rb#L12-L31

ref; https://github.com/rails/importmap-rails?tab=readme-ov-file#using-npm-packages-via-javascript-cdns

---

> Can we generate this automatically when we deploy our application?

Yes, we can generate this in the following ways.

1. We always call the `bin/importmap pin packages` each package because there is no commands to install all dependencies now.

```console
$ bin/import pin package-name
```

2. We always refer to the packages provided at CDN, such as the following.

It was changed because each package had to be requested individually, leading to cost problems for each request.

```diff
diff --git a/config/importmap.rb b/config/importmap.rb
index 9fb275b..69ffe84 100644
--- a/config/importmap.rb
+++ b/config/importmap.rb
@@ -2,6 +2,6 @@

 pin "application"
 pin "@rails/actioncable", to: "actioncable.esm.js"
-pin "@rails/ujs", to: "@rails--ujs.js" # @7.0.8
+pin "@rails/ujs", to: "https://ga.jspm.io/npm:@rails/ujs@7.0.8-4/lib/assets/compiled/rails-ujs.js"
```

kou replied:
Hmm. I tried `rails new todo` with Rails 7.1.3.4 and `vendor/javascript/` had only `.keep`.
Can we remove `@rails/ujs` dependency?

FYI: Here is the `config/importmap.rb` generated by `rails new todo`:

```rb
# Pin npm packages by running ./bin/importmap

pin "application"
pin "@hotwired/turbo-rails", to: "turbo.min.js"
pin "@hotwired/stimulus", to: "stimulus.min.js"
pin "@hotwired/stimulus-loading", to: "stimulus-loading.js"
pin_all_from "app/javascript/controllers", under: "controllers"
```

kou replied:
Here is the `config/import.rb` generated by `rails new todo --skip-hotwire`:

```rb
# Pin npm packages by running ./bin/importmap

pin "application"
```

otegami replied:
fix: [b6ca351](https://github.com/ranguba/ranguba/pull/41/commits/b6ca3512a56e0798f809b6c859c6978c61434685)
Sure. I just removed the `@rails/ujs` dependency.
I also didn't add Hotwire because we didn't use `@rails/ujs`'s function in the current Ranguba, although I thought we had to migrate `@rails/ujs` to Hotwire/Turbo.

otegami replied:
In addition, we don't use Action Cable in Ranguba. If it's okay, I will want to remove Action Cable in the following PR.

otegami committed:
bin/importmap unpin @rails/ujs
